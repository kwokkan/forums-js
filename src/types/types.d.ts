
interface IForumSessionUser {
    id: number;
    name: string;
    joinedDate: number;
}

interface INextSession extends any {
    forumsUser?: IForumSessionUser;
}

interface INextProvider extends any {
    id: string;
    name: string;
}

declare module "next-auth" {
    interface INextAccount extends any {
        provider: string;
        id: string;
    }

    export interface IProviderOptions {
        providers: any[];
        callbacks?: IProviderOptionsCallback;
        events?: INextEventsOptions;
        pages?: INextPageOptions;
        debug?: boolean;
    }

    interface IProviderOptionsCallback {
        signIn?: (user: INextUser, account: INextAccount, profile: INextProfile) => Promise<bool>;
        redirect?: (url: string, baseUrl: string) => Promise<string>;
        session?: (session: INextSession, user: INextUser, sessionToken: string) => Promise<INextSession>;
        jwt?: (token: INextToken, user: INextUser, account: INextAccount, profile: INextProfile, isNewUser: boolean) => Promise<INextToken>;
    }

    interface INextEventsOptions {
        signIn?: (message: any) => Promise;
        signOut?: (message: any) => Promise;
        createUser?: (message: any) => Promise;
        linkAccount?: (message: any) => Promise;
        session?: (message: any) => Promise;
        error?: (message: any) => Promise;
    }

    interface INextPageOptions {
        signIn?: string;
        signOut?: string;
        error?: string;
        verifyRequest?: string;
        newUser?: string;
    }

    interface INextProfile extends any {
    }


    interface INextToken extends any {
        forumsUser?: IForumSessionUser;
    }

    interface INextUser extends any {
        name: string;
        forumsUser?: IForumSessionUser;
    }

    function handler(req: IncomingMessage, res: ServerResponse, options: IProviderOptions): void;

    export = handler;
}

declare module "next-auth/client" {
    export const Provider: any;

    export function getSession(context: { req: IncomingMessage }): Promise<INextSession>;

    export function providers<T>(context: GetServerSidePropsContext<T>): Promise<INextProvider[]>;

    export function signIn(provider?: string, args?: { callbackUrl?: string }): void;

    export function signOut(): void;

    export function useSession(): [INextSession | null, boolean];
}

declare module "next-auth/providers" {
    interface IProviderConfig {
        clientId: string,
        clientSecret?: string,
        scope?: string[] | string
    }

    export = {
        GitHub: (config: IProviderConfig) => any
    };
}
