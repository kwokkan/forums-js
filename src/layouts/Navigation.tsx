import Link from "next/link";
import React from "react";
import { Icon, Nav, Navbar } from "rsuite";

interface IProps {
    title: string;
}

export function Navigation(props: IProps) {
    return (
        <Navbar>
            <Navbar.Header>
                <Link href="/">
                    <a className="navbar-brand logo">
                        <img className="mr-3" src="/logo.png" title={props.title} alt="Logo" width="24" height="24" />
                        {props.title}
                    </a>
                </Link>
            </Navbar.Header>

            <Navbar.Body>
                <Nav>
                    <Link href="/" passHref>
                        <Nav.Item eventKey="1" icon={<Icon icon="home" />}>
                            Home
                        </Nav.Item>
                    </Link>

                    <Nav.Item eventKey="2">Forums</Nav.Item>
                    <Nav.Item eventKey="3">New Posts</Nav.Item>

                    <Link href="/about" passHref>
                        <Nav.Item eventKey="4">
                            About
                        </Nav.Item>
                    </Link>
                </Nav>

                <Nav pullRight>
                    <Nav.Item icon={<Icon icon="github" />} href="https://github.com/kwokkan/forums-js" target="_blank" rel="noreferrer noopener">
                        Fork me
                    </Nav.Item>
                    <Link href="/settings" passHref>
                        <Nav.Item icon={<Icon icon="cog" />}>Settings</Nav.Item>
                    </Link>
                </Nav>
            </Navbar.Body>
        </Navbar>
    );
}
