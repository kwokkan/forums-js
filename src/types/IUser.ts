export interface IUser {
    id: number;
    name: string;
    joinedDate: Date;
}
