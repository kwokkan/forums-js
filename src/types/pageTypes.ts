import { GetServerSideProps } from "next";
import { ParsedUrlQuery } from "querystring";

export type GetTypedServerSideProps<TParams extends ParsedUrlQuery, TProps = { [key: string]: any }> = GetServerSideProps<TProps, TParams>;
