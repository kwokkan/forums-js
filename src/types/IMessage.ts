import { IUser } from "./IUser";

export interface IMessage {
    id: number;
    created: Date;
    content: string;
    user: IUser;
}
