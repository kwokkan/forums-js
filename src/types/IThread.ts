import { IMessage } from "./IMessage";

export interface IThread {
    id: number;
    created: number;
    name: string;
    messages: IMessage[];
}
