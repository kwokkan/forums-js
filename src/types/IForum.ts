import { IThread } from "./IThread";

export interface IForum {
    id: number;
    created: Date;
    name: string;
    description?: string;
    threads?: IThread[];
}
