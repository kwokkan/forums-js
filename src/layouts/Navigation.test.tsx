/**
 * @jest-environment jsdom
*/
import React from "react";
import { create } from "react-test-renderer";
import { IUser } from "../types/IUser";
import { Navigation } from "./Navigation";

test("Renders with title", () => {
    const tree = create(
        <Navigation title="Test app" />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

test("Renders with user", () => {
    const user: IUser = {
        id: 1,
        name: "Test user",
        joinedDate: 1
    };

    const tree = create(
        <Navigation title="Test app" user={user} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
