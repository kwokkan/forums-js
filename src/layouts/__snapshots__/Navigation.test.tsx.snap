// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with title 1`] = `
<nav
  className="rs-navbar rs-navbar-default"
>
  <a
    className="rs-navbar-brand"
  >
    <img
      alt="Logo"
      className="mr-3"
      height="24"
      src="/logo.png"
      title="Test app"
      width="24"
    />
    Test app
  </a>
  <div
    className="rs-navbar-nav rs-nav rs-nav-default rs-nav-horizontal"
  >
    <a
      className="rs-navbar-item"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <svg
        aria-hidden={true}
        aria-label="explore"
        className="rs-icon"
        data-category="application"
        fill="currentColor"
        focusable={false}
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
      >
        <path
          d="M8 15A7 7 0 108 1a7 7 0 000 14zm0 1A8 8 0 118 0a8 8 0 010 16z"
        />
        <path
          d="M9.238 9.238l1.858-4.334-4.334 1.858 2.477 2.477zM6 6l5.851-2.508a.5.5 0 01.657.657L10 10l-5.851 2.508a.5.5 0 01-.657-.657L6 6z"
        />
      </svg>
      Home
      <span
        className="rs-ripple-pond"
      >
        <span
          className="rs-ripple"
        />
      </span>
    </a>
    <a
      className="rs-navbar-item"
      href="#"
      onClick={[Function]}
      role="button"
    >
      Forums
      <span
        className="rs-ripple-pond"
      >
        <span
          className="rs-ripple"
        />
      </span>
    </a>
    <a
      className="rs-navbar-item"
      href="#"
      onClick={[Function]}
      role="button"
    >
      New Posts
      <span
        className="rs-ripple-pond"
      >
        <span
          className="rs-ripple"
        />
      </span>
    </a>
    <a
      className="rs-navbar-item"
      href="/about"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      About
      <span
        className="rs-ripple-pond"
      >
        <span
          className="rs-ripple"
        />
      </span>
    </a>
  </div>
  <div
    className="rs-navbar-nav rs-navbar-right rs-nav rs-nav-default rs-nav-horizontal"
  >
    <a
      className="rs-navbar-item"
      href="https://github.com/kwokkan/forums-js"
      onClick={[Function]}
      rel="noreferrer noopener"
      target="_blank"
    >
      <svg
        aria-hidden={true}
        aria-label="rate"
        className="rs-icon"
        data-category="application"
        fill="currentColor"
        focusable={false}
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
      >
        <path
          d="M2.58 2.995L2 2.582l-.58.413a.042.042 0 01-.034.006.04.04 0 01-.027-.049l.205-.695-.56-.435a.06.06 0 01-.023-.046.06.06 0 01.058-.061L1.74 1.7l.225-.664a.037.037 0 01.023-.023.037.037 0 01.047.023l.225.664.701.015a.06.06 0 01.036.106l-.56.435.205.695a.04.04 0 01-.006.034.039.039 0 01-.055.009zM2 3.809a1.04 1.04 0 001.6-1.141l-.012-.04.021-.016A1.06 1.06 0 002.981.715a1.036 1.036 0 00-1.963 0 1.06 1.06 0 00-.627 1.896l.021.016-.012.04a1.04 1.04 0 001.601 1.142zm.58 5.186L2 8.581l-.58.413A.042.042 0 011.386 9a.04.04 0 01-.027-.049l.205-.695-.56-.435a.06.06 0 01-.023-.046.06.06 0 01.058-.061l.701-.015.225-.664a.037.037 0 01.023-.023.037.037 0 01.047.023l.225.664.701.015a.06.06 0 01.036.106l-.56.435.205.695a.04.04 0 01-.006.034.039.039 0 01-.055.009zM2 9.809a1.04 1.04 0 001.6-1.141l-.012-.04.021-.016a1.06 1.06 0 00-.628-1.897 1.037 1.037 0 00-1.963 0 1.06 1.06 0 00-.627 1.896l.021.016-.012.04a1.04 1.04 0 001.601 1.142zm.58 5.186L2 14.581l-.58.413a.042.042 0 01-.034.006.04.04 0 01-.027-.049l.205-.695-.56-.435a.06.06 0 01-.023-.046.06.06 0 01.058-.061l.701-.015.225-.664a.037.037 0 01.023-.023.037.037 0 01.047.023l.225.664.701.015a.06.06 0 01.036.106l-.56.435.205.695a.04.04 0 01-.006.034.039.039 0 01-.055.009zm-.58.814a1.04 1.04 0 001.6-1.141l-.012-.04.021-.016a1.06 1.06 0 00-.628-1.897 1.037 1.037 0 00-1.963 0 1.06 1.06 0 00-.627 1.896l.021.016-.012.04a1.04 1.04 0 001.601 1.142zM8.58 2.995L8 2.582l-.58.413a.042.042 0 01-.034.006.04.04 0 01-.027-.049l.205-.695-.56-.435a.06.06 0 01-.023-.046.06.06 0 01.058-.061L7.74 1.7l.225-.664a.037.037 0 01.023-.023.037.037 0 01.047.023l.225.664.701.015a.06.06 0 01.035.106l-.56.435.205.695a.04.04 0 01-.006.034.039.039 0 01-.055.009zM8 3.809a1.04 1.04 0 001.6-1.141l-.012-.04.021-.016A1.06 1.06 0 008.981.715a1.036 1.036 0 00-1.964 0 1.06 1.06 0 00-.627 1.896l.021.016-.012.04A1.04 1.04 0 008 3.809zm.58 5.186L8 8.581l-.58.413A.042.042 0 017.386 9a.04.04 0 01-.027-.049l.205-.695-.56-.435a.06.06 0 01-.023-.046.06.06 0 01.058-.061l.701-.015.225-.664a.037.037 0 01.023-.023.037.037 0 01.047.023l.225.664.701.015a.06.06 0 01.035.106l-.56.435.205.695a.04.04 0 01-.006.034.039.039 0 01-.055.009zM8 9.809a1.04 1.04 0 001.6-1.141l-.012-.04.021-.016a1.06 1.06 0 00-.628-1.897 1.037 1.037 0 00-1.964 0 1.06 1.06 0 00-.627 1.896l.021.016-.012.04A1.04 1.04 0 008 9.809zM14.58 2.995L14 2.582l-.58.413a.042.042 0 01-.034.006.04.04 0 01-.027-.049l.205-.695-.56-.435a.06.06 0 01-.023-.046.06.06 0 01.058-.061l.701-.015.225-.664a.037.037 0 01.023-.023.037.037 0 01.047.023l.225.664.701.015a.06.06 0 01.035.106l-.56.435.205.695a.04.04 0 01-.006.034.039.039 0 01-.055.009zm-.58.814a1.04 1.04 0 001.6-1.141l-.012-.04.021-.016a1.06 1.06 0 00-.628-1.897 1.036 1.036 0 00-1.964 0 1.06 1.06 0 00-.627 1.896l.021.016-.012.04A1.04 1.04 0 0014 3.809z"
        />
      </svg>
      Fork me
      <span
        className="rs-ripple-pond"
      >
        <span
          className="rs-ripple"
        />
      </span>
    </a>
    <a
      className="rs-navbar-item"
      href="/api/auth/signin"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <svg
        aria-hidden={true}
        aria-label="gear"
        className="rs-icon"
        data-category="action"
        fill="currentColor"
        focusable={false}
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
      >
        <path
          d="M11.967 4.96l-.354-.461.845-2.002-1.935-1.117-1.315 1.738-.574-.073a4.986 4.986 0 00-1.28.003l-.576.076-1.319-1.743-1.935 1.117.85 2.015-.35.461a5.078 5.078 0 00-.638 1.106l-.224.534-2.17.271v2.234l2.169.271.223.537c.164.396.378.769.638 1.11l.35.461-.847 2.008 1.935 1.117 1.312-1.733.576.076c.428.056.862.056 1.291.001l.576-.075 1.31 1.731 1.935-1.117-.842-1.996.352-.461a5.052 5.052 0 00.65-1.125l.223-.536 2.15-.269V6.885l-2.154-.269-.223-.536a4.983 4.983 0 00-.646-1.118zm1.569.733l1.579.197a1 1 0 01.876.992v2.234a1 1 0 01-.876.992l-1.574.197a6.127 6.127 0 01-.778 1.348l.617 1.461a1 1 0 01-.421 1.255l-1.935 1.117a1 1 0 01-1.297-.262l-.96-1.268a5.994 5.994 0 01-1.549-.001l-.961 1.269a.999.999 0 01-1.297.262l-1.935-1.117a1 1 0 01-.421-1.255l.622-1.475a5.974 5.974 0 01-.765-1.332l-1.592-.199a1 1 0 01-.876-.992V6.882a1 1 0 01.876-.992l1.596-.199a6.156 6.156 0 01.764-1.324l-.625-1.482a1 1 0 01.421-1.255L4.96.513a1 1 0 011.297.262l.968 1.279a5.982 5.982 0 011.536-.003L9.727.775a.999.999 0 011.297-.262l1.935 1.117a1 1 0 01.421 1.255l-.619 1.467c.318.414.577.866.775 1.341z"
        />
        <path
          d="M8 10a2 2 0 10.001-3.999A2 2 0 008 10zm0 1a3 3 0 110-6 3 3 0 010 6z"
        />
      </svg>
      Login
      <span
        className="rs-ripple-pond"
      >
        <span
          className="rs-ripple"
        />
      </span>
    </a>
  </div>
</nav>
`;

exports[`Renders with user 1`] = `
<div
  className="rs-navbar rs-navbar-default"
  role="navigation"
>
  <div
    className="rs-navbar-header"
  >
    <a
      className="navbar-brand logo"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <img
        alt="Logo"
        className="mr-3"
        height="24"
        src="/logo.png"
        title="Test app"
        width="24"
      />
      Test app
    </a>
  </div>
  <div
    className="rs-navbar-body"
  >
    <div
      className="rs-nav rs-nav-default rs-navbar-nav rs-nav-horizontal"
    >
      <ul>
        <li
          className="rs-nav-item"
        >
          <a
            className="rs-nav-item-content"
            href="/"
            onClick={[Function]}
            onMouseEnter={[Function]}
            role="button"
            tabIndex={0}
          >
            <i
              className="rs-icon rs-icon-home"
            />
            Home
            <span
              className="rs-ripple-pond"
            >
              <span
                className="rs-ripple"
                style={Object {}}
              />
            </span>
          </a>
        </li>
        <li
          className="rs-nav-item"
        >
          <a
            className="rs-nav-item-content"
            onClick={[Function]}
            role="button"
            tabIndex={0}
          >
            Forums
            <span
              className="rs-ripple-pond"
            >
              <span
                className="rs-ripple"
                style={Object {}}
              />
            </span>
          </a>
        </li>
        <li
          className="rs-nav-item"
        >
          <a
            className="rs-nav-item-content"
            onClick={[Function]}
            role="button"
            tabIndex={0}
          >
            New Posts
            <span
              className="rs-ripple-pond"
            >
              <span
                className="rs-ripple"
                style={Object {}}
              />
            </span>
          </a>
        </li>
        <li
          className="rs-nav-item"
        >
          <a
            className="rs-nav-item-content"
            href="/about"
            onClick={[Function]}
            onMouseEnter={[Function]}
            role="button"
            tabIndex={0}
          >
            About
            <span
              className="rs-ripple-pond"
            >
              <span
                className="rs-ripple"
                style={Object {}}
              />
            </span>
          </a>
        </li>
      </ul>
    </div>
    <div
      className="rs-nav rs-nav-default rs-navbar-nav rs-navbar-right rs-nav-horizontal"
    >
      <ul>
        <li
          className="rs-nav-item"
        >
          <a
            className="rs-nav-item-content"
            href="https://github.com/kwokkan/forums-js"
            onClick={[Function]}
            rel="noreferrer noopener"
            role="button"
            tabIndex={0}
            target="_blank"
          >
            <i
              className="rs-icon rs-icon-github"
            />
            Fork me
            <span
              className="rs-ripple-pond"
            >
              <span
                className="rs-ripple"
                style={Object {}}
              />
            </span>
          </a>
        </li>
        <li
          className="rs-nav-item"
        >
          <a
            className="rs-nav-item-content"
            href="/settings"
            onClick={[Function]}
            onMouseEnter={[Function]}
            role="button"
            tabIndex={0}
          >
            <i
              className="rs-icon rs-icon-cog"
            />
            Settings
            <span
              className="rs-ripple-pond"
            >
              <span
                className="rs-ripple"
                style={Object {}}
              />
            </span>
          </a>
        </li>
        <li
          className="rs-nav-item"
        >
          <a
            className="rs-nav-item-content"
            onClick={[Function]}
            role="button"
            tabIndex={0}
          >
            <i
              className="rs-icon rs-icon-cog"
            />
            Logout
            <span
              className="rs-ripple-pond"
            >
              <span
                className="rs-ripple"
                style={Object {}}
              />
            </span>
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
